unobstructed_lanechange/src/unobstructed_lanechange.cpp

```diff
@@ -0,0 +1,614 @@
+        
+        auto downsampled_points = carma_utils::containers::downsample_vector(points_and_target_speeds, downsample_ratio_);
+
+        cav_msgs::TrajectoryPlan trajectory;
+        trajectory.header.frame_id = "map";
+        trajectory.header.stamp = ros::Time::now();
+        trajectory.trajectory_id = boost::uuids::to_string(boost::uuids::random_generator()());
+
+        trajectory.trajectory_points = compose_trajectory_from_centerline(downsampled_points, req.vehicle_state);
+
+        trajectory.initial_longitudinal_velocity = std::max(req.vehicle_state.longitudinal_vel, minimum_speed_);
+
+        resp.trajectory_plan = trajectory;
+
+        for (int i=0; i<req.maneuver_plan.maneuvers.size(); i++){
+            if (req.maneuver_plan.maneuvers[i].type == cav_msgs::Maneuver::LANE_CHANGE){
+            resp.related_maneuvers.push_back(i);
+            break;
+            }
+        }
+        resp.maneuver_status.push_back(cav_srvs::PlanTrajectory::Response::MANEUVER_IN_PROGRESS);
+
+        return true;
+
+    }
+
+    std::vector<PointSpeedPair> UnobstructedLaneChangePlugin::maneuvers_to_points(const std::vector<cav_msgs::Maneuver>& maneuvers,
+    double max_starting_downtrack, const carma_wm::WorldModelConstPtr& wm,const cav_msgs::VehicleState& state){
+        std::vector<PointSpeedPair> points_and_target_speeds;
+        std::unordered_set<lanelet::Id> visited_lanelets;
+
+        bool first = true;
+        for(const auto& maneuver : maneuvers)
+        {
+            if(maneuver.type != cav_msgs::Maneuver::LANE_CHANGE)
+            {
+                throw std::invalid_argument("Unobstructed Lane Change does  not support this maneuver type");
+            }
+            cav_msgs::LaneChangeManeuver lane_change_maneuver = maneuver.lane_change_maneuver;
+            
+            double starting_downtrack = lane_change_maneuver.start_dist;
+            if(first){
+                if(starting_downtrack > max_starting_downtrack)
+                {
+                    starting_downtrack = max_starting_downtrack;
+                }
+                first = false;
+            }
+            // //Get lane change route
+            double ending_downtrack = lane_change_maneuver.end_dist;
+                //Get geometry for maneuver
+            if (starting_downtrack >= ending_downtrack)
+            {
+                throw std::invalid_argument("Start distance is greater than or equal to end distance");
+            }
+
+            //get route geometry between starting and ending downtracks
+            lanelet::BasicLineString2d route_geometry = create_route_geom(starting_downtrack, stoi(lane_change_maneuver.starting_lane_id), ending_downtrack, wm);
+            
+            int nearest_pt_index = getNearestRouteIndex(route_geometry,state);
+            lanelet::BasicLineString2d future_route_geometry(route_geometry.begin() + nearest_pt_index, route_geometry.end());
+
+            first = true;
```

Comment:
This is rather confusing using nested `first`s or resetting this variable every loop. Is there a more straightforward way to express this? If not, let's rename or add comments for clarity